### DISCUSSION TITLE
Documentation unclear: Are `when` blocks necessary?

### DISCUSSION THREAD

**Discussion by neuschaefer** [woodpecker-ci/woodpecker#3774] [Ask a Question] [answered]
I am currently struggling to get a woodpecker workflow going, and I think my problems are related to `when` blocks.

The [documentation](https://woodpecker-ci.org/docs/next/usage/workflow-syntax#when---conditional-execution) states:

> when - Conditional Execution
> Woodpecker supports defining a list of conditions for a step by using a when block. [...]

My naive interpretation is:

- `when` is for _conditional_ execution. I don't need to specify it for _unconditional_ execution (run this workflow on every suitable event)

But I suspect I am wrong about this, because I get a "bad habit" warning when I don't specify a `when` block, and for some reason which I have not fully investigated, my workflows aren't automatically triggered by a push.

Related but different issue: https://github.com/woodpecker-ci/woodpecker/issues/3773

---

The second thing that I'm unclear about is whether when blocks can be specified per workflow file, or only inside a step. As far as I understand the examples in the documentation, both are possible.

**Comment by zc-devs** [comment, marked as answer]
> when is for conditional execution. I don't need to specify it for unconditional execution

You are right. Worked 5 days ago https://github.com/woodpecker-ci/woodpecker/pull/3655#issuecomment-2145827975.

> workflows aren't automatically triggered by a push

Check if webhook works.

> whether when blocks can be specified per workflow file, or only inside a step. As far as I understand the examples in the documentation, both are possible.

> both are possible

[when - Global workflow conditions](https://woodpecker-ci.org/docs/usage/workflow-syntax#when---global-workflow-conditions)
[when - Conditional Execution](https://woodpecker-ci.org/docs/usage/workflow-syntax#when---conditional-execution)

**Reply by neuschaefer** [reply]
Thank you for your clarifications!

I'm not sure how to check if webhooks work, because I don't run this woodpecker instance (it is https://ci.codeberg.org)

**Reply by qwerty287** [reply]
> I'm not sure how to check if webhooks work

Check the webhook settings at codeberg.

Note that there's a bug currently if you enabled the repo recently you have to use `Repair repository` after enabling to get it working.

**Reply by neuschaefer** [reply]
After clicking "repair repository", builds are indeed triggered by push.

Unfortunately I can't really provide more information, but my problems are solved now. Thank you for your patience!

**Reply by qwerty287** [reply]
That was a bug on our end, but it's fixed in 2.6.0

**Comment by qwerty287** [comment]
> But I suspect I am wrong about this, because I get a "bad habit" warning when I don't specify a when block

You should filter your pipelines by event because we may add additional events in the future and  you probably don't want to run your pipelines on every of them. See https://woodpecker-ci.org/docs/usage/linter#event-filter-for-all-steps

**Reply by neuschaefer** [reply]
Thank you. It's starting to make sense.
